File: typescript/test/useCase/OrderCreationUseCase.test.ts
import Category from '../../src/domain/Category';
import Order from '../../src/domain/Order';
import { OrderStatus } from '../../src/domain/OrderStatus';
import Product from '../../src/domain/Product';
import { ProductCatalog } from '../../src/repository/ProductCatalog';
import OrderCreationUseCase from '../../src/useCase/OrderCreationUseCase';
import SellItemRequest from '../../src/useCase/SellItemRequest';
import SellItemsRequest from '../../src/useCase/SellItemsRequest';
import UnknownProductException from '../../src/useCase/UnknownProductException';
import InMemoryProductCatalog from '../doubles/InMemoryProductCatalog';
import TestOrderRepository from '../doubles/TestOrderRepository';

describe('OrderApprovalUseCase', () => {
  const orderRepository: TestOrderRepository = new TestOrderRepository();
  let food: Category = new Category();
  food.setName('food');
  food.setTaxPercentage(10);

  const saladProduct = new Product();
  saladProduct.setName('salad');
  saladProduct.setPrice(3.56);
  saladProduct.setCategory(food);
  const tomatoProduct = new Product();
  tomatoProduct.setName('tomato');
  tomatoProduct.setPrice(4.65);
  tomatoProduct.setCategory(food);
  const productCatalog: ProductCatalog = new InMemoryProductCatalog([ saladProduct, tomatoProduct]);
  const useCase: OrderCreationUseCase = new OrderCreationUseCase(orderRepository, productCatalog);

  it('sellMultipleItems', () => {
      let saladRequest: SellItemRequest = new SellItemRequest();
      saladRequest.setProductName('salad');
      saladRequest.setQuantity(2);

      let tomatoRequest: SellItemRequest = new SellItemRequest();
      tomatoRequest.setProductName('tomato');
      tomatoRequest.setQuantity(3);

      let request: SellItemsRequest = new SellItemsRequest();
      request.setRequests([]);
      request.getRequests().push(saladRequest);
      request.getRequests().push(tomatoRequest);

      useCase.run(request);

      const insertedOrder: Order = orderRepository.getSavedOrder();
      expect(insertedOrder.getStatus()).toBe(OrderStatus.CREATED);
      expect(insertedOrder.getTotal()).toBe(23.20);
      expect(insertedOrder.getTax()).toBe((2.13));
      expect(insertedOrder.getCurrency()).toBe(('EUR'));
      expect(insertedOrder.getItems().length).toBe(2);
      expect(insertedOrder.getItems()[0].getProduct().getName()).toBe('salad');
      expect(insertedOrder.getItems()[0].getProduct().getPrice()).toBe(3.56);
      expect(insertedOrder.getItems()[0].getQuantity()).toBe(2);
      expect(insertedOrder.getItems()[0].getTaxedAmount()).toBe(7.84);
      expect(insertedOrder.getItems()[0].getTax()).toBe(0.72);
      expect(insertedOrder.getItems()[1].getProduct().getName()).toBe('tomato');
      expect(insertedOrder.getItems()[1].getProduct().getPrice()).toBe(4.65);
      expect(insertedOrder.getItems()[1].getQuantity()).toBe(3);
      expect(insertedOrder.getItems()[1].getTaxedAmount()).toBe(15.36);
      expect(insertedOrder.getItems()[1].getTax()).toBe(1.41);
  });

  it('unknownProduct', () => {
      let request: SellItemsRequest = new SellItemsRequest();
      request.setRequests([]);
      let unknownProductRequest: SellItemRequest = new SellItemRequest();
      unknownProductRequest.setProductName('unknown product');
      request.getRequests().push(unknownProductRequest);

      expect(() => useCase.run(request)).toThrow(UnknownProductException);
  });
});

File: scala/src/test/scala/usecase/OrderCreationUseCaseTest.scala
package usecase

import doubles.{InMemoryProductCatalog, TestOrderRepository}
import ordershipping.domain.{Category, OrderStatus, Product}
import ordershipping.usecase.{
  OrderCreationUseCase,
  SellItemRequest,
  SellItemsRequest,
  UnknownProductException
}
import org.scalatest.BeforeAndAfterEach
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

class OrderCreationUseCaseTest
    extends AnyFlatSpec
    with Matchers
    with BeforeAndAfterEach {
  private val food = new Category(name = "food", taxPercentage = 10)
  private val productCatalog = new InMemoryProductCatalog(
    List(
      new Product(name = "salad", price = 3.56, category = food),
      new Product(name = "tomato", price = 4.65, category = food)
    )
  )
  private var orderRepository: TestOrderRepository = _
  private var useCase: OrderCreationUseCase = _

  override def beforeEach(): Unit = {
    orderRepository = new TestOrderRepository()
    useCase = new OrderCreationUseCase(
      orderRepository = orderRepository,
      productCatalog = productCatalog
    )
  }

  "order creation use case" should "sell multiple items" in {
    val saladRequest = SellItemRequest(productName = "salad", quantity = 2)
    val tomatoRequest = SellItemRequest(productName = "tomato", quantity = 3)

    val request = SellItemsRequest(
      List(saladRequest, tomatoRequest)
    )

    useCase.run(request)

    val insertedOrder = orderRepository.insertedOrder
    insertedOrder.status shouldBe OrderStatus.Created
    insertedOrder.total shouldBe 23.20
    insertedOrder.tax shouldBe 2.13
    insertedOrder.currency shouldBe "EUR"
    insertedOrder.items.length shouldBe 2
    insertedOrder.items.head.product.name shouldBe "salad"
    insertedOrder.items.head.product.price shouldBe 3.56
    insertedOrder.items.head.quantity shouldBe 2
    insertedOrder.items.head.taxedAmount shouldBe 7.84
    insertedOrder.items.head.tax shouldBe 0.72

    insertedOrder.items(1).product.name shouldBe "tomato"
    insertedOrder.items(1).product.price shouldBe 4.65
    insertedOrder.items(1).quantity shouldBe 3
    insertedOrder.items(1).taxedAmount shouldBe 15.36
    insertedOrder.items(1).tax shouldBe 1.41
  }

  "order creation use case" should "unknown product" in {
    val request = SellItemsRequest(
      List(SellItemRequest(productName = "unknown product", quantity = 0))
    )
    assertThrows[UnknownProductException] {
      useCase.run(request)
    }
  }
}
